### PR TITLE
Improved code coverage for "info.limpet.data.operations" package

### PR DESCRIPTION
**Pull request by kumarbabu**
AddLayerOperation.java

### CODE REVIEW

File: info.limpet.test/src/info/limpet/data/TestOperations.java [hunk 1]
```diff
@@ -800,6 +803,39 @@ public void testSubtraction()
...
+    InMemoryStore store = new SampleData().getData(10);
+
+    List<IStoreItem> selection = new ArrayList<IStoreItem>();
+
+    StoreGroup track1 = new StoreGroup("Track 1");
+    selection.add(track1);
+
+    Collection<ICommand<IStoreItem>> commands =
+        new AddLayerOperation().actionsFor(selection, store, context);
+    assertEquals("invalid number of inputs", 1, commands.size());
```
Comment (IanMayo) [inline info.limpet.test/src/info/limpet/data/TestOperations.java]: hmm, I think the message for this test is that it is proving the "valid number of commands", not an invalid number of inputs.

File: info.limpet.test/src/info/limpet/data/TestOperations.java [hunk 2]
```diff
@@ -800,6 +803,39 @@ public void testSubtraction()
...
+    for (ICommand<IStoreItem> iCommand : commands)
+    {
+      iCommand.execute();
+    }
+
```
Comment (IanMayo) [inline info.limpet.test/src/info/limpet/data/TestOperations.java:951]: the above command creates a new layer.  I think you should test that a layer actually got added into the StoreGroup.

File: info.limpet.test/src/info/limpet/data/TestOperations.java [hunk 3]
```diff
@@ -800,6 +803,39 @@ public void testSubtraction()
...
+    selection.add(speedGood1);
+
+    commands = new AddLayerOperation().actionsFor(selection, store, context);
+    assertEquals("invalid number of inputs", 1, commands.size());
+    for (ICommand<IStoreItem> iCommand : commands)
+    {
+      iCommand.execute();
+      iCommand.dataChanged(speedGood1);
+    }
+  }
```
Comment (IanMayo) [inline info.limpet.test/src/info/limpet/data/TestOperations.java:1081]: from looking at the AddLayerOperation class, I can see that if there isn't anything provided in the selection, then a new layer is created at the top level of the store.

**Comment by IanMayo** [comment]
@kumarbabu - have updated the branch to match the current master branch. 

I believe you have a couple more steps to do.

But, before you do, please note that the application has been repackaged, so there are some new projects, and some have changed name.

When you do a "Pull", you may need to import some of the new projects from the git folder - so Eclipse knows about the new ones.

Back to the changes.

You've made a good attempt.  But, as we would expect, your understanding of Limpet is holding you back from fully testing the class, IMHO.  You "didn't turn up" for our chat last Wednesday afternoon - maybe we should try again.  

I've also inserted some inline comments to your PR, to help you along :-)

**Comment by kumarbabu** [comment]
I have updated the local repository with your guidelines. After performing the PULL operation I have found new projects in GIT/limpet folder and I have imported them as a eclipse projects into local workspace. However after building the workspace I found following error showing in the IStoreGroup.java class.

ERROR: Duplicate methods named spliterator with the parameters () and () are inherited from the types Collection<IStoreItem> and Iterable<IStoreItem>
I am using JDK8, Java7 compiler and eclipse3.8.2

Work around:
I have fixed the issue with following configurations.
JDK 7, Java 7 compiler and eclipse 3.8.2

**Comment by IanMayo** [comment]
Ooh, was trying to get online to give you the workaround (JRE7) :-)

I see you've now fixed it. Well done.

**Comment by kumarbabu** [comment]
Hi,
After PULL the latest code, then I have performed the maven build with "mvn package" in limpet root directory. The following errors are occurred  during the package build operations.

> > > mvn package
> > > [INFO] Scanning for projects...
> > > [ERROR] Internal error: java.lang.RuntimeException: Could not resolve target platform specification artifact info.limpet:eclipse38
> > > :target:1.0.6-SNAPSHOT -> [Help 1]
> > > org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: Could not resolve target platform specificati
> > > on artifact info.limpet:eclipse38:target:1.0.6-SNAPSHOT
> > >         at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:168)
> > >         at org.apache.maven.cli.MavenCli.execute(MavenCli.java:537)
> > >         at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:196)
> > >         at org.apache.maven.cli.MavenCli.main(MavenCli.java:141)
> > >         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> > >         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> > >         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> > >         at java.lang.reflect.Method.invoke(Method.java:483)
> > >         at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:290)
> > >         at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:230)
> > >         at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:409)
> > >         at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:352)
> > > Caused by: java.lang.RuntimeException: Could not resolve target platform specification artifact info.limpet:eclipse38:target:1.0.6
> > > -SNAPSHOT
> > >         at org.eclipse.tycho.core.resolver.DefaultTargetPlatformConfigurationReader.addTargetArtifact(DefaultTargetPlatformConfigu
> > > rationReader.java:389)
> > >         at org.eclipse.tycho.core.resolver.DefaultTargetPlatformConfigurationReader.setTarget(DefaultTargetPlatformConfigurationRe
> > > ader.java:342)
> > >         at org.eclipse.tycho.core.resolver.DefaultTargetPlatformConfigurationReader.getTargetPlatformConfiguration(DefaultTargetPl
> > > atformConfigurationReader.java:75)
> > >         at org.eclipse.tycho.core.resolver.DefaultTychoResolver.setupProject(DefaultTychoResolver.java:86)
> > >         at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.afterProjectsRead(TychoMavenLifecycleParticipant.java:90)
> > >         at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:274)
> > >         at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
> > >         ... 11 more
> > > [ERROR]
> > > [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> > > [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> > > [ERROR]
> > > [ERROR] For more information about the errors and possible solutions, please read the following articles:
> > > [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/InternalErrorException

Could you please help me to resolve this issue. 

Here I am trying to check the code coverage improvements. Please suggest me, am I in right track to "mvn package".

Regards,
Kumar.

**Comment by pecko** [comment]
You should run the following once after changing target platform (it changes rarely; just once until now)

```
cd info.limpet.targetplatforms
mvn clean install
```

**Comment by kumarbabu** [comment]
Before code fixes.
before

After code fixes.
after

**Comment by kumarbabu** [comment]
Please find the code coverage improvement.

after2
after1

**Comment by kumarbabu** [comment]
Improved code coverage for few more classes in info.limpet.data.operations.admin package.

**Before:**

before_operations_admin
before_limpet_on27thfeb

**After Code fixes:**
after_operation_admin
after3

**Comment by kumarbabu** [comment]
few more code coverage improvements...

after_above 80

**Pull request merged**